### ISSUE TITLE
GPX Export

### ISSUE THREAD

**Issue by jojorb** [gabm/FancyPlaces#7]
Hi thank for you app,

i'm looking forward to find the "geo files" you use to store the list of fancy places.
where is it store inside the phone?
can you tell me if it's geojson or gpx friendly or any other format?

i am very interested to be able to "save them" on a personal nodejs webapp i'm currently working on

**Comment by gabm** [comment]
well currently everything except the images are stored in a sqlite database - but i was already thinking about some kind of webservice - what we could do for sure would be an exporter to what ever format you like (probably gpx) .. maybe also a "real" interface but that would be quite some work...

**Comment by jojorb** [comment]
a GPX exporter could be awesome, i do believe that users would want at some points be able to save their places somewhere outside the phone. :)

**Comment by gabm** [comment]
I am sure about that and I agree with you - lets start with a gpx exporter :)

**Comment by jojorb** [comment]
Great news, i will follow that ;)

**Comment by gabm** [comment]
Hey Roby, please try out the new Debug Version that supports a very simple gpx export... you can download it [here](https://www-user.tu-chemnitz.de/~gabm/FancyPlaces/FancyPlaces.apk)

Currently the functionality is very limited: you can only export one place at a time and the filename is fixed... but its already prepared for exporting multiple files, currently we need support for multiple selection of the FancyPlaces in the List, so first we need Issue #8 to be fixed...

Please test with your webapp if the gpx is valid (i tried with an online tool for now) and report back...

**Comment by jojorb** [comment]
the export goes well on phone,
when i did plug my phone (let's say i don't share by BT or Wifi, but by USB)
i had hard time to find the Fancy Places file
so i copy past the .gpx to my /phone/download file to be able to see it on my computer.
the .gpx on webbapp seems ok but if i may it could be nice to have it like this:

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<gpx version="1.1"
creator="GT-I9300"
xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
xmlns="http://www.topografix.com/GPX/1/1"
xsi:schemaLocation="http://www.topografix.com/GPX/1/1 http://www.topografix.com/GPX/1/1/gpx.xsd">
    <metadata>
        <name>test</name>
        <time>2015-06-08T14:14:54Z</time>
    </metadata>
    <wpt lat="48.9476478" lon="2.5227681">
    </wpt>
</gpx>
```

if the phone model can be as Creator or whatever something meaning full about phone,
could be awesome!

**Comment by jojorb** [comment]
i just re think about something you may already figure out but i put it there

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<gpx version="1.1"
creator="FANCY-PLACE"
xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
xmlns="http://www.topografix.com/GPX/1/1"
xsi:schemaLocation="http://www.topografix.com/GPX/1/1 http://www.topografix.com/GPX/1/1/gpx.xsd">
<metadata>
<author><name>GT-I9300</name></author>
<name>test</name>
<time>2015-06-08T14:14:54Z</time>
</metadata><trk>
...
```

**Comment by k3b** [comment]
i have tested gpx and found a minor issue:

your app is setting time to local time while time="...Z" means utc. so you can either onmint tze "Z" or translate the time to gmt/utc

you can also add 

<pre>
    &lt;name>theName&lt;/name>
    &lt;desc>the Notes&lt;/desc>
    &lt;time>2015-02-10T08:04:45.000Z&lt;/time>
    &lt;link href='file://Path/to/Image.jpg' /> &lt;!-- or a "content:..." uri -->
</pre>

**Comment by gabm** [comment]
thank you both for your feedback - i will implement your thoughts...

I'm not sure whether a link to an uri on the local device would be very helpful here, so maybe i'll drop that...

**Comment by k3b** [comment]
@RobyRemzy i think the name and time element should not belong to the meta element but to the wpt element as already implemented. this way the gpx-file can possibly contain more than one fance place.

**Comment by k3b** [comment]
@gabm you are right: a link to a local-android file does not make sense if the gpx file is copied somewhere to an other computer. 

if the gpx file is viewed on the originating android-devjce it makes sense. A portable way would be to copy the image as a file to the same dir as the gpx file.

I plan that the next version of the https://f-droid.org/wiki/page/de.k3b.android.locationMapViewer / https://github.com/k3b/LocationMapViewer will support images when loading gpx files.

But I agree: the link to the image is the least important info in the gpx and can be onmitted

**Comment by gabm** [comment]
@k3b well then it would make sense, indeed.. is there any standard regarding images and gpx together? like a simple zip archive or something like that?

Probably images and gpx in one directory would make the most sense - if the user decides to just send the gpx, its his own decision...

**Comment by jojorb** [comment]
@k3b yes you probably right, more convenient out of metatag. 
i'm more used to work with .gpx with tracking files not single point.
and i think it's cool to have the pics inside the same folder.
maybe i don't know if you can push to get also elevation and speed? for very fancy photo! :) 
    `<ele>55</ele>`
    `<speed>3.6770509481430054</speed>`

**Comment by gabm** [comment]
Hi guys,

i am currently preparing the v1.2 release which will officially feature the gpx export... Could u please test [this](https://www-user.tu-chemnitz.de/~gabm/FancyPlaces/FP-20150621-1758.apk) test build again and report back?

GPX Export works now like this:
- a folder with the current date is created in sdcard: Fancy Places/Export_TIMESTAMP
- within this folder you can find a file "export.gpx" and also the corresponding image files, if they are available...
- Currently i dont want to implement a directory picker because it seems quite some work (?) ... this would make things more convenient for sure...

Well please report back if the feature works for you...

@k3b Relative paths in URIs are not standardised...How to you handle this in your app? It seems its widely accepted...

**Comment by jojorb** [comment]
Hi there,
thank you for the update,
for now i don't mind too much to have a folder for each "poi".
i just had some difficulties to fetch the folder when my phone is plug on my computer, for any reasons the folder don't appear,  it's probably due to my ROM, i dunno.

in the folder i have my export.gpx and my 0_mytest.png
there it could be awesome to have the same name for the .gpx and the png.

also inside the .gpx i can see that
name : my test
desc : my test (witch should be : this is my note)

**Comment by gabm** [comment]
about the folder acess on your phone: "i don't know"

the rest is working as expected i think:

since you now can select and export multiple FancyPlaces at once the policy is like this:
- 1 folder per export
- 1 gpx file for ALL FancyPlaces in the export folder
- 1 image file for every FancyPlace in the gpx file (if a photo has been provided)

it seems there is a copy+paste error and the description field contains the title currently, right?

**Comment by jojorb** [comment]
oh i see it's 1 folder for selected POI and 1 export.gpx. 

and yes the title is also into the description field, right now.

**Comment by gabm** [comment]
okay, the bug should be fixed now, a new test release is [here](https://www-user.tu-chemnitz.de/~gabm/FancyPlaces/FP-20150621-2213.apk)

@RobyRemzy can u import the gpx to your app? Does everything work for you?

**Comment by k3b** [comment]
@gabm uri with relative paths:

that is a good question.

the map-symbol can be interpretet as a path relative to the gpx-url.

this way the gpx can be loaded via file or via http:
- http://my.server.com/path/myGpx.gpx" with a symbol "images/myImage.jpg" means http://my.server.com/path/images/myImage.jpg
- file://c:/path/myGpx.gpx" with a symbol "images/myImage.jpg" means  file://c:/path/images/myImage.jpg

**Comment by gabm** [comment]
thats interesting - I thought I read somewhere, that "relative URIs" dont exist.. but english [wikipedia](https://en.wikipedia.org/wiki/Uniform_resource_identifier) says differently...

Currently relative filenames are already exported.. maybe you could test it (see my previous comment) and if it works I'll close here for now and prepare the v1.2.0 release...

**Comment by k3b** [comment]
i have tested interoperability with FP-20150621-2213.apk

potential problems in fp are formatted as **bold**

usecase filemanager open gpx-file
- LocationMapViewer opens map with the FancyPoints. when clicking on bubble i get text and description from gpx. **gpx-link** "file://2_test.png" is currently interpreted as link. clicking on it crashes the app because "2_test.png" was not found in the root folder . (i have to fix it on LocationMapViewer (https://github.com/k3b/LocationMapViewer/issues/8 and https://github.com/k3b/LocationMapViewer/issues/9). I assume the path should be either "file:2_test.png" (without the "//") or "2_test.png"
- c:geo (geocaching app) open FancyPoints as cache-locations. only reads name and location but not description or link (because it uses gpx-1.0 format while fancy-places uses gpx-1.1)
- OsmAnd~ (Navigation app) opens FancyPoints with name and description (no link)
- **open a non-fp-gpx-file with FancyPlaces**: fp starts and nothing seems to happen: no error-message/feedback. gpx file seems not to be interpreted at all 

usecase in FancyPlaces share a fancy place
- **geo-uri does not use url-encoding** to replace special chars if the are no special chars then it seems ok. Example if you have name= romeo&julia then this must be encoded as romeo%26julia
- LocationMapViewer ok for name and location. (description is not working probably a bug in LocationMapViewer (https://github.com/k3b/LocationMapViewer/issues/7))
- OsmAnd~ (Navigation app) ok for name and location.

Example how to urlencode fp-name

```
  String urlEnecodedName = java.net.URLEncoder.encode(fp.getName(), "UTF-8");
```

**Comment by gabm** [comment]
Hi, thank you a lot for your extensive testing!
- **gpx-link**
  - I suppose, that "file:2_test.png" would be a good idea, but honestly: i couldn't find any "standard" here
- **open a non-fp-gpx-file with FancyPlaces**
  - actually nothing is supposed to happen but i have to remove FancyPlaces registering for opening files
  - Gpx Import is not implemented yet, i am still thinking about a "good" exchange format... GPX ist extensible but cannot transport images and i dont want users to send "directories of gpx and images" ...
- **geo-uri does not use url-encoding**
  - thought I did that, but i guess there is a bug here - I am investigating this!

**Comment by k3b** [comment]
```
 > gpx-link
```

I donot think that there is a standard for this, too- i prefer "2_test.png"  because that would also allow tu use http: https: urls. else we download gpx via http that references a file.

But that decision is up to you. my codeupdate for k3b/LocationMapViewer#9 will do it the way you will do.

```
 > fp export
```

you can use the LocationMapViewer-s parser. it already supports gpx-1.0, gpx-1.1, kml format and handles different time formats (gmt-time, local-time, time with timezone, ...) so it should be possible to import geodata form other apps, too.

For details see

https://github.com/k3b/LocationMapViewer/blob/master/k3b-geoHelper/src/test/java/de/k3b/geo/io/gpx/GpxReaderTest.java (this is an android independant j2se-junittest)

if you are lucky all you have to do is replace 

```
   import de.k3b.geo.api.GeoPointDto;
   import de.k3b.geo.api.IGeoPointInfo;
```

with you code.

IGeoPointInfo is the data interface and GeoPointDto a simple implementation of IGeoPointInfo 

---

As export-format you can do the same as openoffice or google-kmz-geo-format:
 a zip file containing all files (ie gpx+*.png)

**Comment by gabm** [comment]
ok, so now things should be alright for now, you can try it [here](https://www-user.tu-chemnitz.de/~gabm/FancyPlaces/FP-20150623-1740.apk)
- dropped the "//" in the file link... so for now we go with: "file:2_test.png" ... 
- fixed the uri encoding when viewing a location in a mapping application

I will now prepare release v1.2.0 ... The next steps will include work on a working fancy place ex- and import (i like your idea of the zip file).. Maybe I can use your gpx reader and maybe we could also think about a gpx writer implementation?!

Thank you both for your help and your time!

**Comment by k3b** [comment]
I just checked the New testbuild and found a bug in gpx export. There is no XML escaping 

I Exporte a FP named "Romeo&Juliet" and See that the "&" is not xml-escaped (&amp;amp;)

I have no Permission to reopen
 this ticket

**Comment by gabm** [comment]
I added a very simple and stupid method to fix this issue for now - I need to build a proper XML reader/writer, probably using org.sax library (and maybe i can reuse some of your code)...

just for the record: the issue is fixed using this [method](http://www.leveluplunch.com/java/examples/escape-xml-characters/) which maybe doesn't even care about all cases...

Its included in v1.2.0 ...
